### ISSUE TITLE
Dead Zone Adjustment

### ISSUE THREAD

**Issue by Spidey002** [InsanityBringer/PiccuEngine#70]
Need a way to adjust the dead zone for controllers. I have a huge dead zone on all axes that make precise aiming difficult.

**Comment by Lukrative525** [comment]
I second this. The large dead zone makes precise aiming difficult when using a flight stick.

**Comment by Glowhyena** [comment]
> I second this. The large dead zone makes precise aiming difficult when using a flight stick.

I agree. It made me want to stop playing the game.

**Comment by InsanityBringer** [comment]
Does running with -deadzone 0 in your command line parameters help? in any case it probably should be more visible. I think I could fit it into the controls menu maybe.

**Comment by Lukrative525** [comment]
> Does running with -deadzone 0 in your command line parameters help?

I just tested this, and as far as I can tell, it doesn't work. I tracked the lowest axis values that induced rotation with and without the command line argument, and they were the same. Is there a deadzone value stored in my .plt file that is overriding the command line argument?

**Comment by Spidey002** [comment]
> I think I could fit it into the controls menu maybe.

That would be awesome!

**Comment by InsanityBringer** [comment]
My bad, can you try adding `-deadzone0 0 -deadzone1 0`? (this is a problem, you can only set the deadzone for two devices.. frig I hate the Descent 3 input system). I seem to have misremembered what the parameters were.

**Comment by Lukrative525** [comment]
@InsanityBringer Thanks, I tried that out and that does seem to improve things. However, it still doesn't reduce the deadzone all the way, and I think this section of code might be to blame:

https://github.com/InsanityBringer/PiccuEngine/blob/56cab84f088e08bea8b69bf230e2923b9af6b508/win32/wincontroller.cpp#L683

**Comment by Glowhyena** [comment]
> [@InsanityBringer](https://github.com/InsanityBringer) Thanks, I tried that out and that does seem to improve things. However, it still doesn't reduce the deadzone all the way, and I think this section of code might be to blame:
> 
> [PiccuEngine/win32/wincontroller.cpp](https://github.com/InsanityBringer/PiccuEngine/blob/56cab84f088e08bea8b69bf230e2923b9af6b508/win32/wincontroller.cpp#L683)
> 
> Line 683 in [56cab84](/InsanityBringer/PiccuEngine/commit/56cab84f088e08bea8b69bf230e2923b9af6b508)
>  nullzone = (m_ControlList[controller].deadzone < 0.05f) ? 0.05f : m_ControlList[controller].deadzone;

Yeah, I experienced it. ;/

**Comment by InsanityBringer** [comment]
should there be no minimum deadzone at all? that would be easy to do though I don't know if the somewhat noisy readings will cause you to drift. since it would be slow

**Comment by Lukrative525** [comment]
Yes, I think there should be no minimum drift. Unfortunately, when I tried it out, I found that the math done on the input from my stick (which outputs values in 0-65535 with 32767 as its center) results in asymmetric normalized values biased in the negative direction, so I get negative drift in all axes when the stick is centered.

**Comment by Lukrative525** [comment]
It took me an embarrassingly long time to work out how the axis values are getting modified as they change hands getting from ddio_win/winjoy.cpp/joymm_get_pos to win32/wincontroller.cpp/gameWinController::get_axis_value, but as a test, I removed all modifications of the axis values that happen before gameWinController::get_axis_value and set deadzone/nullzone to zero (see 89d6efbe2e6e4fd785a2e0caf2c21000d0734c20). Obviously this was done as an experiment with no regard for how these changes might affect data retrieved using XInput or any other API.

But doing that makes sure that the axis values visible to gameWinController::get_axis_value are the same ones reported by WinMM. In addition to functionally deleting joy_GetRawPos, this removes the negative drift bias and also improves the feel somewhat, but it's clear that WinMM is reporting much lower resolution than that provided by my stick over HID, meaning there is still a de facto deadzone caused by the low resolution of the axis values.

So it seems that something like 89d6efbe2e6e4fd785a2e0caf2c21000d0734c20 is going to be the best we can do using WinMM, but it would have to be implemented in a way that doesn't break data gathered from other joystick API's. It looks like an SDL3 implementation is in the works, so maybe I'll mess around with that.
